### PR TITLE
Refactor and add merge vis.js config for nodes and edges

### PR DESCRIPTION
**Pull request by dukesun99**
As according to #144, a new config field named `advanced` can be added, in which the config fields will be auto merged as vis.js config for nodes and edges. Previous config fields are kept but refactored.
Example of usage as follows:
```js
"Character": {
	"caption": "name",
	"size": "pagerank",
	"community": "community",
	//"sizeCypher": "MATCH (n) WHERE id(n) = {id} MATCH (n)-[r]-() RETURN sum(r.pagerank) AS c",
        "advanced": {
		//"borderWidth": {value: 10, type:0},
                //"borderWidth": {value: "community", type:1},
                "borderWidth": {value: "match (n) where id(n) = $id return n.community", type:2},
                 //"borderWidth": {value: (node) => {return node.properties.community}, type:3},
         }
}
```
Current issues:

> This version of neo4j lacks automatic type conversion, i.e. previously using int type as node label is acceptable (implemented in #113), this pr rollback this feature. Maybe we need to consider adding it in the future or think of a way to let users specify explicit conversion.

### CODE REVIEW

**Comment by thebestnom** [comment]
Hmm... I think my point is that I want to completely merge configs (yes I do mean completely break api)

So things like `ILabel` should extend/have inside the original Label config from vis-network in the d.ts

**Comment by thebestnom** [comment]
I see something like
```javascript
{
   //...
   labels: { testLabel: {
      normal: {image:"url"//or anothet name}
      func: {size: props => props.test}
      cypher: { anotherLabrlProp: "cypher"}
   }}
}
```
Or something of that kind

**Comment by thebestnom** [comment]
The problem with `advance` is what eill happened if someone want advance as a prop

**Comment by dukesun99** [comment]
> The problem with `advance` is what eill happened if someone want advance as a prop

Using `advanced` itself should not have problems functionally because the allowed properties are defined by us, such as `caption`, `sizeCypher`. If we do not want to use `advanced` in the future, it should be ok. Inside `advanced` if there is another `advanced` it is perfectly fine.  



> Hmm... I think my point is that I want to completely merge configs (yes I do mean completely break api)
> 
> So things like `ILabel` should extend/have inside the original Label config from vis-network in the d.ts

Yes in my way ts is not working... That is a problem.



> I see something like
> 
> ```js
> {
>    //...
>    labels: { testLabel: {
>       normal: {image:"url"},//or anothet name
>       func: {size: props => props.test},
>       cypher: { anotherLabrlProp: "cypher"}
>    }}
> }
> ```
> 
> Or something of that kind
> 
> And then every prop can be any of them

This is doable but it would completely break the API, and also personally I feel for not so technical people, it is harder for them to read the vis.js API than just using the current API, which is just follow the example without truly understanding how things work. So I feel it is meaningful to keep the simpler version of API for such people.

**Comment by dukesun99** [comment]
@thebestnom What do you think? Should we change the API completely?

**Comment by thebestnom** [comment]
Maybe you can add the advance the same way I added default, what do you say? But I do want to at least break the api to be exactly the same as vis (in the d.ts just extend the Node and Relationship of vis-network), then we can put example and some documentation ourself and then link to vis configuration for mote advance use cases

**Comment by dukesun99** [comment]
@thebestnom You mean break the api by, for example, changing `community` to `group` which is a vis config field?

**Comment by dukesun99** [comment]
@thebestnom What about like this? Then we can let the static part be exactly the same as vis api
```js
labels: {
    //"Character": "name",
    'Character': {
        'caption': 'name',
        'size': 'pagerank',
        'community': 'community',
        //"sizeCypher": "MATCH (n) WHERE id(n) = {id} MATCH (n)-[r]-() RETURN sum(r.weight) AS c"
        [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
            [NeoVis.NEOVIS_STATIC]: {
                'borderWidth': 10,
                'anotherone': 'something'
            },
            [NeoVis.NEOVIS_PEOPERTY_NAME]: {
                'borderWidth': 'pagerank'
            },
            [NeoVis.NEOVIS_CYPHER]: {
                'borderWidth': 'match (n) where id(n) = $id return n.community'
            },
            [NeoVis.NEOVIS_FUNCTION]: {
                'borderWidth': (node) => {
                    return node.properties.community;
                }
            },
        }
    }
},
```

**Comment by thebestnom** [comment]
> @thebestnom You mean break the api by, for example, changing `community` to `group` which is a vis config field? Or only for advanced part we follow vis api exactly?

Exactly

And for the second one, after `[​NeoVis​.​NEOVIS_ADVANCED_CONFIG​]​:` there is no reason to use symbols again as we know that its our object so
```javascript
{
[​NeoVis​.​NEOVIS_ADVANCED_CONFIG​]​: ​{​
           // ​[​NeoVis​.​NEOVIS_STATIC​]​: ​{​ not needed, already handled by breaking the api
              //  ​'borderWidth'​: ​10​,​
           // //    ​'anotherone'​: ​'something'​
           // ​}​,​
            ​propertyName: ​{​
                ​'borderWidth'​: ​'pagerank'​
            ​}​,​
            ​cypher: ​{​
                ​'borderWidth'​: ​'match (n) where id(n) = $id return n.community'​
            ​}​,​
            ​function: ​{​
                ​'borderWidth'​: ​(​node​)​ ​=>​ ​{​
                    ​return​ ​node​.​properties​.​community​;​
                ​}​
            ​}​,​
        ​}​
}
```
Is enough

**Comment by dukesun99** [comment]
@thebestnom ok. The only problem is we will lose backward compatibility completely. I will modify the code and perhaps need your help with d.ts as I am not familiar with it.

**Comment by dukesun99** [comment]
@thebestnom Besides, I feel there are still some issues:

1. For `title_properties`, I feel it is better to keep it as it is convenient to use. 
2. some vis.js config fields for nodes and edges are objects. Basically, we need a way to let user construct objects using cypher or some other neater ways.

**Comment by dukesun99** [comment]
@thebestnom Now the usage is like this:
```js
labels: {
    //"Character": "name",
    'Character': {
        'shape': 'dot',
        //"sizeCypher": "MATCH (n) WHERE id(n) = {id} MATCH (n)-[r]-() RETURN sum(r.weight) AS c"
        [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
            'propertyName': {
                'value': 'pagerank',
                'group': 'community',
                'label': 'name'
            },
            'function': {
                'title': (node) => {
                    return viz.propertyNamesToString(node, undefined);
                }
            },
        }
    }
},
relationships: {
    'INTERACTS': {
        //"sizeCypher": "MATCH (n) WHERE id(n) = {id} MATCH (n)-[r]-() RETURN sum(r.weight) AS c"
        [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
            'propertyName': {
                'value': 'weight',
            },
            'function': {
                'title': (edge) => {
                    return viz.propertyNamesToString(edge, undefined);
                }
            },
        }
    }
},
```
The above code performs the same as simple-example.html.

**Comment by thebestnom** [comment]
looks great! Ill look at the actual code later

> some vis.js config fields for nodes and edges are objects. Basically, we need a way to let user construct objects using cypher or some other neater ways.

I think that allowing nested object here would be easier for the user instead of object build from cypher, but it does make the code harder

**Comment by thebestnom** [comment]
Also I meant the exact opisite, static is not handled, `propertyName` is already handled, my bad, sorry for the misinterpretting
how you did it last time was coorect just not needed `propertyName` in the advance config

**Comment by dukesun99** [comment]
> Also I meant the exact opisite, static is not handled, `propertyName` is already handled, my bad, sorry for the misinterpretting
> how you did it last time was coorect just not needed `propertyName` in the advance config

You mean make propertyName the default outside advanced config and static inside?

**Comment by dukesun99** [comment]
> looks great! Ill look at the actual code later
> 
> > some vis.js config fields for nodes and edges are objects. Basically, we need a way to let user construct objects using cypher or some other neater ways.
> 
> I think that allowing nested object here would be easier for the user instead of object build from cypher, but it does make the code harder

Ok sure that would be doable, I will design that when I have time.

**Comment by thebestnom** [comment]
Yes, that mostly how it was before

**Comment by dukesun99** [comment]
> Yes, that mostly how it was before

Yeah sure.

**Comment by dukesun99** [comment]
Current sample usage is as below:
```js
'Character': {
    'value': 'pagerank',
    'group': 'community',
    'label': 'name',
    //"sizeCypher": "MATCH (n) WHERE id(n) = {id} MATCH (n)-[r]-() RETURN sum(r.weight) AS c"
    [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
        'static': {
            'shape': 'dot',
        },
        'function': {
            'title': (node) => {
                return viz.propertyNamesToString(node, undefined);
            }
        },
    }
}
},
relationships: {
'INTERACTS': {
    'value': 'weight',
    //"sizeCypher": "MATCH (n) WHERE id(n) = {id} MATCH (n)-[r]-() RETURN sum(r.weight) AS c"
    [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
        'function': {
            'title': (edge) => {
                return viz.propertyNamesToString(edge, undefined);
            }
        },
    }
}
},
```

**Comment by dukesun99** [comment]
@thebestnom What do you think about this approach?
```js
'Character': {
    'value': 'pagerank',
    'group': 'community',
    'label': 'name',
    'font': {
        'size':'pagerank'
    }
    [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
        'static': {
            'shape': 'dot',
            'font': {
                'color': '#343434'
            }
        },
        'function': {
            'title': (node) => {
                return viz.propertyNamesToString(node, undefined);
            }
        },
    }
}
},
```

**Comment by thebestnom** [comment]
That's looks great! Though why do we still need `viz​.​propertyNamesToString`?

**Comment by dukesun99** [comment]
> propertyNamesToString

`propertyNamesToString` is the helper function to assemble the HTML code for tooltips (captions).

**Comment by dukesun99** [comment]
> nodeToHtml

Ok I'll change that.

**Comment by dukesun99** [comment]
@thebestnom Ok I have implemented object as a config. The config objects will be merged if have multiple occurrence.   
I still need your help with d.ts thing. Could you help to have a look when you have time?

**Comment by dukesun99** [comment]
Unit test is also lacking btw... I will update it tomorrow

**Comment by dukesun99** [comment]
@thebestnom OK the unit tests are updated. Only problem is the d.ts. I'll flag it as ready for review and maybe you can help to update it?

**Comment by thebestnom** [comment]
Sure, Im trying to do it with mapped types so it would be 100% matched from network-viz

**Comment by thebestnom** [comment]
Ok, added `d.ts` in the most complicated way I could (now the type is exactly what vis is getting) also I allowed the function type to return object if needed, even though I didn't looked yet if you had allowed that to do so

Havn't looked at everything yet, but looks good so far! only thinking of we should split `neovis.js` to multiple files because it starting to get big

**Comment by dukesun99** [comment]
> Ok, added `d.ts` in the most complicated way I could (now the type is exactly what vis is getting) also I allowed the function type to return object if needed, even though I didn't looked yet if you had allowed that to do so
> 
Great! Thanks!
> 
> Havn't looked at everything yet, but looks good so far! only thinking of we should split `neovis.js` to multiple files because it starting to get big

NeoVis currently is one class right? Maybe we move some helper functions out?

File: src/neovis.js [hunk 1]
```diff
@@ -108,178 +110,171 @@ export default class NeoVis {
...
+		return results;
+	}
 
-		node.id = neo4jNode.identity.toInt();
-		node.raw = neo4jNode;
+	_runFunction(func, node) {
+		if (typeof func === 'function') {
+			return func(node);
+		}
+		return undefined;
```
Comment (thebestnom) [inline src/neovis.js:141]: I think throwing an error is better here

File: src/neovis.js [hunk 2]
```diff
@@ -108,178 +110,171 @@ export default class NeoVis {
...
-				node.value = sizeProp;
-			} else if (sizeProp && typeof sizeProp === 'object' && Neo4j.isInt(sizeProp)) {
-				// property value might be a Neo4j Integer, check if we can call toNumber on it:
-				if (sizeProp.inSafeRange()) {
-					node.value = sizeProp.toNumber();
+	_retrieveProperty(prop, node) {
+		if (typeof node === 'object' && typeof node.properties === 'object') {
+			return node.properties[prop];
+		}
+		return undefined;
```
Comment (thebestnom) [inline src/neovis.js:148]: Same as `_runFunction`

File: src/neovis.js [hunk 3]
```diff
@@ -108,178 +110,171 @@ export default class NeoVis {
...
-			edge.value = r.properties[weightKey];
-		} else if (weightKey && typeof weightKey === 'number') {
-			edge.value = weightKey;
-		} else {
-			edge.value = 1.0;
-		}
+		this._buildPropertyNameObject(nodeTypeConfig, edge, r);
 
-		// set caption
+		if (advancedConfig && typeof advancedConfig === 'object') {
```
Comment (thebestnom) [inline src/neovis.js:269]: throw in case of non object

Comment (thebestnom) [inline src/neovis.js:269]: throw in case of non object

**Comment by thebestnom** [comment]
Ok, finally had time to look at your code, looks great, some small change requet

Also I opened a branch for 2.0.0 for breaking changes and this going to be merged to there

**Comment by dukesun99** [comment]
@thebestnom Ok thanks! I changed the code to throw some errors. Let me know if you think we should define some error classes instead of just throwing strings.

**Comment by thebestnom** [comment]
Didn't got notification for some reason, just throw the error as `new Error("error")`

**Comment by dukesun99** [comment]
@thebestnom Done

**Comment by thebestnom** [comment]
Thank ☺️!

**Pull request merged**